File: model/code_names.go
package model

import (
	"go/token"
	"slices"
	"sort"
	"strconv"
	"strings"
)

// ValidCodeName 检查可直接用于生成符号的树内代码名，不接受空值、Go 关键字或非 ASCII 字符。
func ValidCodeName(name string) bool {
	if len(name) == 0 || len(name) > 40 || !asciiLetter(name[0]) || token.IsKeyword(name) {
		return false
	}
	for i := 1; i < len(name); i++ {
		c := name[i]
		if !asciiLetter(c) && c != '_' && (c < '0' || c > '9') {
			return false
		}
	}
	return true
}

// asciiLetter 判断单字节是否为英文字母。
func asciiLetter(c byte) bool { return c >= 'a' && c <= 'z' || c >= 'A' && c <= 'Z' }

// opaqueNodeID 识别编辑器生成的随机身份，避免将 UUID 再复制进可读代码名。
func opaqueNodeID(id string) bool {
	if len(id) == 37 && strings.EqualFold(id[:5], "node_") {
		id = id[5:]
	} else if len(id) == 36 && id[8] == '-' && id[13] == '-' && id[18] == '-' && id[23] == '-' {
		id = strings.ReplaceAll(id, "-", "")
	} else {
		return false
	}
	for i := 0; i < len(id); i++ {
		c := id[i]
		if (c < '0' || c > '9') && (c < 'a' || c > 'f') && (c < 'A' || c > 'F') {
			return false
		}
	}
	return true
}

// defaultCodeName 优先复用简短语义 ID；随机或非法 ID 则返回类型基名，并要求追加数字。
func defaultCodeName(n Node) (string, bool) {
	name := n.ID
	forceNumber := !ValidCodeName(name) || opaqueNodeID(name)
	if forceNumber {
		name = "Node"
		if n.Type.Valid() {
			name = n.Type.String()
		}
	}
	if name[0] >= 'a' && name[0] <= 'z' {
		name = string(name[0]-'a'+'A') + name[1:]
	}
	return name, forceNumber
}

// availableCodeName 为碰撞名称分配数字后缀，同一截短前缀共享游标，避免反复扫描已占用名称。
// 类型回退通过 forceNumber 从 1 开始分配，不占用无后缀的语义名称。
func availableCodeName(base string, forceNumber bool, used map[string]struct{}, next map[string]int) string {
	if _, exists := used[base]; !exists && !forceNumber {
		return base
	}
	for digits, lower := 1, 1; ; digits, lower = digits+1, lower*10 {
		prefix := base[:min(len(base), 40-digits)]
		key := strconv.Itoa(digits) + ":" + prefix
		start := max(lower, next[key])
		for value := start; value < lower*10; value++ {
			next[key] = value + 1
			candidate := prefix + strconv.Itoa(value)
			if _, exists := used[candidate]; !exists {
				return candidate
			}
		}
	}
}

// WithCodeNames 复制树与节点切片并只补齐缺失代码名，保留显式名称供 Validate 报告错误。
// 按节点 ID 排序分配默认名，因此节点数组重排不会影响首次补全结果。
func WithCodeNames(p Project) Project {
	p.Trees = slices.Clone(p.Trees)
	for treeIndex := range p.Trees {
		tree := &p.Trees[treeIndex]
		tree.Nodes = slices.Clone(tree.Nodes)
		used := make(map[string]struct{}, len(tree.Nodes))
		missing := make([]int, 0)
		for i, node := range tree.Nodes {
			if node.CodeName == "" {
				missing = append(missing, i)
			} else {
				used[node.CodeName] = struct{}{}
			}
		}
		sort.Slice(missing, func(i, j int) bool { return tree.Nodes[missing[i]].ID < tree.Nodes[missing[j]].ID })
		next := make(map[string]int)
		for _, index := range missing {
			node := &tree.Nodes[index]
			base, forceNumber := defaultCodeName(*node)
			node.CodeName = availableCodeName(base, forceNumber, used, next)
			used[node.CodeName] = struct{}{}
		}
	}
	return p
}

File: web/src/codeNames.ts
import type { BTNode, Project, Tree } from "./project.ts";
import { nodeTypes } from "./enums.ts";

// Go 关键字不能作为独立代码名，规则与后端输入校验一致。
const keywords = new Set("break default func interface select case defer go map struct chan else goto package switch const fallthrough if range type continue for import return var".split(" "));

// 代码名限制为短 ASCII 标识符，不修剪输入或隐式修正显式名称。
export function validCodeName(value: unknown): value is string {
  return typeof value === "string" && value.length <= 40
    && /^[A-Za-z]/.test(value) && !/[^A-Za-z0-9_]/.test(value) && !keywords.has(value);
}

// 默认名优先沿用可读短 ID；编辑器自动 UUID 回退到节点类型。
export function defaultCodeName(node: BTNode): { base: string; forceNumber: boolean } {
  const randomID = /^node_[0-9a-f]{32}$/i.test(node.id)
    || /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i.test(node.id);
  const readableID = !randomID && validCodeName(node.id);
  const base = readableID ? node.id : nodeTypes.includes(node.type) ? node.type : "node";
  return { base: base.charAt(0).toUpperCase() + base.slice(1), forceNumber: !readableID };
}

// 占用表和共享后缀游标让批量创建及长名称碰撞的查重保持摊销 O(1)。
export class CodeNameIndex {
  readonly byName = new Map<string, BTNode>(); // 树内显式代码名对应实际节点对象。
  private readonly nextSuffix = new Map<string, number>(); // 同一截短前缀共享数字游标。

  // 建立索引时先保留全部显式值，再按与 Go 一致的 Unicode 顺序补全缺失值。
  constructor(tree: Tree) {
    const missing: { node: BTNode; points: number[] }[] = [];
    for (const node of tree.nodes) {
      if (node.codeName) this.add(node);
      else missing.push({ node, points: Array.from(node.id, (character) => character.codePointAt(0)!) });
    }
    missing.sort((a, b) => {
      for (let i = 0; i < Math.min(a.points.length, b.points.length); i++) {
        if (a.points[i] !== b.points[i]) return a.points[i]! - b.points[i]!;
      }
      return a.points.length - b.points.length;
    });
    for (const { node } of missing) this.add(node);
  }

  // 创建和复制共用分配入口；调用方可传入被复制节点的代码名作为可读基名。
  allocate(base: string, forceNumber = false): string {
    if (!forceNumber && !this.byName.has(base)) return base;
    for (let digits = 1; ; digits++) {
      const prefix = base.slice(0, 40 - digits);
      const key = `${digits}:${prefix}`;
      let suffix = this.nextSuffix.get(key) ?? (digits === 1 ? 1 : 10 ** (digits - 1));
      while (suffix < 10 ** digits) {
        const candidate = `${prefix}${suffix++}`;
        this.nextSuffix.set(key, suffix);
        if (!this.byName.has(candidate)) return candidate;
      }
    }
  }

  // 仅在从业务目录新建节点时取业务函数名；不适合作为代码名时交给类型默认规则。
  // 长函数名截取 40 位后使用同一占用表消歧，名称分配后由节点独立持久保存。
  allocateBusiness(goName: string): string | undefined {
    if (!/^[A-Za-z][A-Za-z0-9_]*$/.test(goName) || keywords.has(goName)) return undefined;
    return this.allocate(goName.slice(0, 40));
  }

  // 复制名称沿用可读基名并分配数字，Wait1 的副本为 Wait2，避免逐次堆叠数字。
  allocateCopy(name: string): string {
    return this.allocate(name.replace(/[0-9]+$/, ""), true);
  }

  // 缺省名称即时写回节点，已有名称校验失败则不改变占用索引。
  add(node: BTNode): void {
    const seed = defaultCodeName(node);
    const name = node.codeName || this.allocate(seed.base, seed.forceNumber);
    if (!validCodeName(name)) throw new Error(`代码名无效：${name}；需为 1–40 位英文开头的字母、数字或下划线，且不能是 Go 关键字`);
    if (this.byName.has(name) && this.byName.get(name) !== node) throw new Error(`当前树内代码名已存在：${name}`);
    node.codeName = name;
    this.byName.set(name, node);
  }

  // 删除不重新编号其他节点；缓存游标保留以避免重复扫描已占用后缀。
  remove(node: BTNode): void {
    if (this.byName.get(node.codeName!) === node) this.byName.delete(node.codeName!);
  }

  // 输入草稿只在应用时进行常数时间格式与唯一性检查。
  validateRename(node: BTNode, name: string): string | null {
    if (!validCodeName(name)) return "代码名需为 1–40 位英文开头的字母、数字或下划线，且不能是 Go 关键字";
    const owner = this.byName.get(name);
    return owner && owner !== node ? `当前树内代码名已存在：${name}` : null;
  }

  // 名称修改只更新节点和占用表，不改 ID、根节点或引用关系。
  rename(node: BTNode, name: string): void {
    const failure = this.validateRename(node, name);
    if (failure) throw new Error(failure);
    this.remove(node);
    node.codeName = name;
    this.add(node);
  }
}

// 未指定代码名的节点进入编辑器或参与语义比较时统一分配，后续保存保留稳定代码名。
export function normalizeCodeNames(project: Project): void {
  for (const tree of project.trees) new CodeNameIndex(tree);
}
